pkg/migration/backup.go

```diff
@@ -0,0 +1,298 @@
+package migration
+
+import (
+	"context"
+	"strings"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// BackupConfigMapName is the name of configmap that contains Daemonset backup.
+	BackupConfigMapName = "px-daemonset-backup"
+)
+
+func (h *Handler) backup(namespace string) error {
+	logrus.Info("Start backup")
+	existingConfigMap := &v1.ConfigMap{}
+	err := h.client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      BackupConfigMapName,
+			Namespace: namespace,
+		},
+		existingConfigMap,
+	)
+	// Only proceed when the configmap does not exist.
+	if !errors.IsNotFound(err) {
+		return err
+	}
+
+	var objs []client.Object
+	if err := h.getDaemonSetComponent(namespace, &objs); err != nil {
+		return err
+	}
+
+	if err := h.getPortworxAPI(namespace, &objs); err != nil {
+		return err
+	}
+	if err := h.getPortworxRbac(namespace, &objs); err != nil {
+		return err
+	}
+
+	if err := h.getStorkComponent(namespace, &objs); err != nil {
+		return err
+	}
+
+	if err := h.getAutopilotComponent(namespace, &objs); err != nil {
+		return err
+	}
+	if err := h.getCSIComponent(namespace, &objs); err != nil {
+		return err
+	}
+	if err := h.getPVCControllerComponent(namespace, &objs); err != nil {
+		return err
+	}
+	if err := h.getMonitoringComponent(namespace, &objs); err != nil {
+		return err
+	}
+
+	var sb strings.Builder
+	for _, obj := range objs {
+		bytes, err := yaml.Marshal(obj)
+		if err != nil {
+			return err
+		}
+
+		sb.WriteString(string(bytes))
+		sb.WriteString("---\n")
+	}
+
+	data := map[string]string{
+		"backup": sb.String(),
+	}
+
+	return h.client.Create(context.TODO(),
```

Comment:
Let's not fail on IsAlreadyPresent error, else on restart the migration will get stuck here. 
Let's print that the backup has already been taken and return nil.

Reply:
this is already checked at the beginning of the function, we already return if configmap exists.

Reply:
Oh sorry, missed that.